File: sdk-php/src/UsageBreakdown.php
<?php

declare(strict_types=1);

namespace Tollara\AgentSdk;

/**
 * Shared usage breakdown for estimate and report responses.
 */
final class UsageBreakdown
{
    public function __construct(
        public ?float $unitsUsed = null,
        public ?float $baseUnitsUsed = null,
        public ?float $overageUnits = null,
        public ?float $chargeableOverageUnits = null,
        public ?float $surplusOverageUnits = null,
        public ?float $overageCost = null,
        public ?float $totalOverageCost = null,
        public ?float $unitsRemaining = null,
        public ?float $remainingCredits = null,
        public ?float $remainingSpendingCap = null,
        public ?float $totalUnitsUsedThisCycle = null,
        public ?bool $overLimit = null,
        public ?bool $overage = null,
        public ?bool $overageAllowed = null,
    ) {
    }

    /** @param array<string, mixed> $data */
    public static function fromArray(array $data): self
    {
        return new self(
            unitsUsed: isset($data['unitsUsed']) ? (float) $data['unitsUsed'] : null,
            baseUnitsUsed: isset($data['baseUnitsUsed']) ? (float) $data['baseUnitsUsed'] : null,
            overageUnits: isset($data['overageUnits']) ? (float) $data['overageUnits'] : null,
            chargeableOverageUnits: isset($data['chargeableOverageUnits']) ? (float) $data['chargeableOverageUnits'] : null,
            surplusOverageUnits: isset($data['surplusOverageUnits']) ? (float) $data['surplusOverageUnits'] : null,
            overageCost: isset($data['overageCost']) ? (float) $data['overageCost'] : null,
            totalOverageCost: isset($data['totalOverageCost']) ? (float) $data['totalOverageCost'] : null,
            unitsRemaining: isset($data['unitsRemaining']) ? (float) $data['unitsRemaining'] : null,
            remainingCredits: isset($data['remainingCredits']) ? (float) $data['remainingCredits'] : null,
            remainingSpendingCap: isset($data['remainingSpendingCap']) ? (float) $data['remainingSpendingCap'] : null,
            totalUnitsUsedThisCycle: isset($data['totalUnitsUsedThisCycle']) ? (float) $data['totalUnitsUsedThisCycle'] : null,
            overLimit: isset($data['isOverLimit']) ? (bool) $data['isOverLimit'] : null,
            overage: isset($data['isOverage']) ? (bool) $data['isOverage'] : null,
            overageAllowed: isset($data['isOverageAllowed']) ? (bool) $data['isOverageAllowed'] : null,
        );
    }
}

File: sdk-python/src/tollara_service_sdk/usage_breakdown.py
from dataclasses import dataclass
from typing import Any, Dict, Optional


@dataclass
class UsageBreakdown:
    units_used: Optional[float] = None
    base_units_used: Optional[float] = None
    overage_units: Optional[float] = None
    chargeable_overage_units: Optional[float] = None
    surplus_overage_units: Optional[float] = None
    overage_cost: Optional[float] = None
    total_overage_cost: Optional[float] = None
    units_remaining: Optional[float] = None
    remaining_credits: Optional[float] = None
    remaining_spending_cap: Optional[float] = None
    total_units_used_this_cycle: Optional[float] = None
    over_limit: Optional[bool] = None
    overage: Optional[bool] = None
    overage_allowed: Optional[bool] = None


def _opt_float(v: Any) -> Optional[float]:
    if v is None:
        return None
    if isinstance(v, (int, float)):
        return float(v)
    return None


def parse_usage_breakdown(data: Optional[Dict[str, Any]]) -> Optional[UsageBreakdown]:
    if data is None or not isinstance(data, dict):
        return None
    return UsageBreakdown(
        units_used=_opt_float(data.get("unitsUsed")),
        base_units_used=_opt_float(data.get("baseUnitsUsed")),
        overage_units=_opt_float(data.get("overageUnits")),
        chargeable_overage_units=_opt_float(data.get("chargeableOverageUnits")),
        surplus_overage_units=_opt_float(data.get("surplusOverageUnits")),
        overage_cost=_opt_float(data.get("overageCost")),
        total_overage_cost=_opt_float(data.get("totalOverageCost")),
        units_remaining=_opt_float(data.get("unitsRemaining")),
        remaining_credits=_opt_float(data.get("remainingCredits")),
        remaining_spending_cap=_opt_float(data.get("remainingSpendingCap")),
        total_units_used_this_cycle=_opt_float(data.get("totalUnitsUsedThisCycle")),
        over_limit=data.get("isOverLimit") if isinstance(data.get("isOverLimit"), bool) else None,
        overage=data.get("isOverage") if isinstance(data.get("isOverage"), bool) else None,
        overage_allowed=data.get("isOverageAllowed")
        if isinstance(data.get("isOverageAllowed"), bool)
        else None,
    )
